### PR TITLE
net: LLDP: Add optional TLV setting and unit tests

### PR DESCRIPTION
**Pull request by finikorg**
On top of PR  #9529

Add setting optional TLV
Add lldp unit tests testing mandatory and optional API

### CODE REVIEW

File: tests/net/lldp/src/main.c [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+static int eth_fake_set_config(struct device *dev,
+			       enum ethernet_config_type type,
+			       const struct ethernet_config *config)
+{
+	struct eth_fake_context *ctx = dev->driver_data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_PROMISC_MODE:
```
Comment (jukkar) [inline tests/net/lldp/src/main.c]: We do not seem to use promiscuous mode flag for anything in this test so the promiscuous mode support be removed

**Comment by nashif** [comment]
what is the status of this?

**Comment by jukkar** [comment]
recheck

File: tests/net/lldp/src/main.c [hunk 2]
```diff
@@ -34,6 +34,7 @@ static struct net_if *iface_posix;
 static struct net_if *iface_fake;
```
Comment (jukkar) [inline tests/net/lldp/src/main.c]: BTW, the commit message of the LLDP tests should start with "tests: net: lldp: ...."

File: samples/net/lldp/src/main.c [hunk 3]
```diff
@@ -17,6 +17,23 @@
 #include <net/net_if.h>
```
Comment (jukkar) [inline samples/net/lldp/src/main.c]: In commit "samples: lldp:....", use "samples: net: lldp: ..." instead.

File: subsys/net/l2/ethernet/lldp/lldp.c [hunk 4]
```diff
@@ -117,10 +116,8 @@ static int lldp_send(struct ethernet_lldp *lldp)
 
 	net_pkt_frag_add(pkt, frag);
 
-	net_buf_add(frag, sizeof(struct net_lldpdu));
-
-	if (!net_pkt_write(pkt, frag, 0, &pos, sizeof(struct net_lldpdu),
-			   (u8_t *)lldp->lldpdu, BUF_ALLOC_TIMEOUT)) {
+	if (!net_pkt_append_all(pkt, sizeof(struct net_lldpdu),
```
Comment (jukkar) [inline subsys/net/l2/ethernet/lldp/lldp.c]: The net_pkt_append_all() is removed, so this needs to use net_pkt_write_new() instead.

**Review by jukkar — CHANGES_REQUESTED** [review]
@finikorg can you fix the merge conflict + other comments, I think we should try to get this merged to 1.14

File: subsys/net/l2/ethernet/lldp/lldp.c (hunk 4 above)
Comment (finikorg) [reply subsys/net/l2/ethernet/lldp/lldp.c]: Updated

File: subsys/net/l2/ethernet/ethernet.c [hunk 5]
```diff
@@ -957,9 +957,32 @@ int net_eth_promisc_mode(struct net_if *iface, bool enable)
 }
 
 #if defined(CONFIG_NET_LLDP)
-int net_eth_set_lldpdu(struct net_if *iface, const struct net_lldpdu *lldpdu)
+static const struct net_lldpdu lldpdu = {
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/ethernet.c]: move that to lldp.c

File: subsys/net/l2/ethernet/ethernet.c [hunk 6]
```diff
@@ -957,9 +957,32 @@ int net_eth_promisc_mode(struct net_if *iface, bool enable)
...
+	},
+#if defined(CONFIG_NET_LLDP_END_LLDPDU_TLV_ENABLED)
+	.end_lldpdu_tlv = NET_LLDP_END_LLDPDU_VALUE
+#endif /* CONFIG_NET_LLDP_END_LLDPDU_TLV_ENABLED */
+};
+
+int net_eth_set_lldpdu(struct net_if *iface)
 {
-	return net_lldp_config(iface, lldpdu);
+	return net_lldp_config(iface, &lldpdu);
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/ethernet.c]: and modify net_lldp_config() so it does not take llpdu as a parameter. For now, there is only a unique llpdu for the system anyway

File: subsys/net/l2/ethernet/lldp/lldp.c [hunk 7]
```diff
@@ -111,6 +111,15 @@ static int lldp_send(struct ethernet_lldp *lldp)
 		goto out;
 	}
 
+	if (lldp->optional_du && lldp->optional_len) {
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/lldp/lldp.c]: you will need modify line 100 the size requested for allocation : basically sizeof(struct net_llpdu) + lldp->optional_len

File: subsys/net/l2/ethernet/lldp/lldp.c [hunk 8]
```diff
@@ -120,6 +120,16 @@ static int lldp_send(struct ethernet_lldp *lldp)
 		}
 	}
 
+#if defined(CONFIG_NET_LLDP_END_LLDPDU_TLV_ENABLED)
+	u16_t tlv_end = htons(NET_LLDP_END_LLDPDU_VALUE);
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/lldp/lldp.c]: never declare a variable in the middle of a code block. Adding { before and } after line 130 could be a solution.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by finikorg** [comment]
Rebased PR and fixed issues noticed in the review.

File: subsys/net/l2/ethernet/lldp/lldp.c [hunk 9]
```diff
@@ -127,6 +127,19 @@ static int lldp_send(struct ethernet_lldp *lldp)
 		}
 	}
 
+#if defined(CONFIG_NET_LLDP_END_LLDPDU_TLV_ENABLED)
```
Comment (tbursztyka) [inline subsys/net/l2/ethernet/lldp/lldp.c]: Use if (IS_ENABLED(CONFIG_NET_LLDP_END_LLDPDU_TLV_ENABLED)) {...} 

It's nicer.

Comment (finikorg) [reply subsys/net/l2/ethernet/lldp/lldp.c]: Yes, looks nicer, updated

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**